prov/shm/src/smr_progress.c

```diff
@@ -842,32 +877,37 @@ static void smr_progress_cmd(struct smr_ep *ep)
 	while (!ofi_cirque_isempty(smr_cmd_queue(ep->region))) {
 		cmd = ofi_cirque_head(smr_cmd_queue(ep->region));
 
-		switch (cmd->msg.hdr.op) {
-		case ofi_op_msg:
-		case ofi_op_tagged:
-			ret = smr_progress_cmd_msg(ep, cmd);
-			break;
-		case ofi_op_write:
-		case ofi_op_read_req:
-			ret = smr_progress_cmd_rma(ep, cmd);
-			break;
-		case ofi_op_write_async:
-		case ofi_op_read_async:
-			ofi_ep_rx_cntr_inc_func(&ep->util_ep, cmd->msg.hdr.op);
-			ofi_cirque_discard(smr_cmd_queue(ep->region));
-			ep->region->cmd_cnt++;
-			break;
-		case ofi_op_atomic:
-		case ofi_op_atomic_fetch:
-		case ofi_op_atomic_compare:
-			ret = smr_progress_cmd_atomic(ep, cmd);
-			break;
-		default:
-			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
-				"unidentified operation type\n");
-			ret = -FI_EINVAL;
+		if (cmd->msg.hdr.op_flags & SMR_CTRL_MSG) {
+			assert(cmd->msg.hdr.op == ofi_ctrl_connreq);
+			smr_progress_connreq(ep, cmd);
+		} else {
```

Comment:
Why can't the op be added to the switch statement, versus checking a flag?  The op_flags should depend on the opcode, not the opcode dependent on a flag.

Reply:
Because there are 2 different sets of ops - ofi_ctrl_*
https://github.com/ofiwg/libfabric/blob/master/include/ofi_proto.h#L52
and ofi_op_*
https://github.com/ofiwg/libfabric/blob/master/include/ofi_proto.h#L121
The op_flags are kind of a misnomer - they really should be just flags or cmd_flags, not op_flags

Reply:
hdr.op *should* be the field that decides how to interpret all other fields.  Having the 5th bit of a flag field be the determining factor is confusing.

I think the best option is to add a base_ctrl_op = ofi_op_max, and have all ctrl ops start there.

Reply:
You mean in ofi_proto.h? Wouldn't that change all the ofi_ctrl_* values and create versioning issues?

Reply:
I'm suggesting that the shm provider define a single op space.  And to limit the changes, that the new ctrl ops use an offset of + ofi_op_max.  IOW, the code here would be:
```
switch (cmd->msg.hdr.op) {
case ofi_op_msg:
case ofi_op_tagged:
    ...
case ofi_op_atomic_compare:
    ...
case ofi_op_max + ofi_ctrl_connreq:
    smr_progress_connreq(ep, cmd)
default:
    FI_WARN(...)
}
```
You can use ofi_op_max directly, or define a new value (base_ctrl_op) for where the ctrl ops should start.  The latter would avoid problems if ofi_op_max is ever changed and could reserve some values for new non-ctrl ops.

Reply:
Btw, if you look at the comment in fi_proto.h
```
 * Intent is that RX can use opcode + control as indices into a function
 * pointer array for message processing (after validating values).
```
This is describing having the basic command opcodes be offset by some number of values reserved for control opcodes.  But since shm didn't have control opcodes before, I would simply reverse the order and place the control opcodes after the basic command opcodes.

The shm protocol version doesn't need to change in this case.  If there's a mismatch, the receiver will simply get a unknown opcode and discard it.

Reply:
Brilliant - thank you